ecosystem/node-checker/src/configuration/types.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+#![allow(dead_code)]
+
+use crate::{metric_evaluator::StateSyncMetricsEvaluatorArgs, runner::BlockingRunnerArgs};
+use anyhow::Result;
+use clap::Parser;
+use once_cell::sync::Lazy;
+use poem_openapi::{types::Example, Object as PoemObject};
+use serde::{Deserialize, Serialize};
+use url::Url;
+
+pub const DEFAULT_METRICS_PORT: u16 = 9101;
+pub const DEFAULT_API_PORT: u16 = 8080;
+pub const DEFAULT_NOISE_PORT: u16 = 6180;
+
+pub static DEFAULT_METRICS_PORT_STR: Lazy<String> =
+    Lazy::new(|| format!("{}", DEFAULT_METRICS_PORT));
+pub static DEFAULT_API_PORT_STR: Lazy<String> = Lazy::new(|| format!("{}", DEFAULT_API_PORT));
+pub static DEFAULT_NOISE_PORT_STR: Lazy<String> = Lazy::new(|| format!("{}", DEFAULT_NOISE_PORT));
+
+// To briefly explain why many of these structs derive 3 different classes of traits:
+// - Parser (clap): To allow users to generate configs easily using nhc configuration create
+// - Serialize / Deserialize (serde): So we can read / write configs from / to disk
+// - PoemoOject: So we can return the configuration over the API
+
+#[derive(Clone, Debug, Deserialize, Parser, PoemObject, Serialize)]
+#[clap(author, version, about, long_about = None)]
+pub struct NodeConfiguration {
+    #[clap(flatten)]
+    pub node_address: NodeAddress,
+
+    /// This is the name we expect clients to send over the wire to select
+    /// which configuration they want to use. e.g. devnet_fullnode
+    #[clap(long)]
+    pub configuration_name: String,
+
+    /// This is the name we will show for this configuration to users.
+    /// For example, if someone opens the NHC frontend, they will see this name
+    /// in a dropdown list of configurations they can test their node against.
+    /// e.g. "Devnet FullNode", "Testnet Validator Node", etc.
+    #[clap(long)]
+    pub configuration_name_pretty: String,
+
+    /// The chain ID we expect to find when we speak to the node.
+    /// If not given, we will just assume the value we find is correct.
+    /// If given, we will check that the value is correct, exiting if not.
+    #[clap(long)]
+    chain_id: Option<u16>,
+
+    /// The role type we expect to find when we speak to the node.
+    /// If not given, we will just assume the value we find is correct.
+    /// If given, we will check that the value is correct, exiting if not.
+    /// e.g. "full_node", "validator_node", etc.
+    #[clap(long)]
+    role_type: Option<String>,
+
+    /// The (metric) evaluators to use, e.g. state_sync, api, etc.
+    #[clap(long, required = true, min_values = 1, use_value_delimiter = true)]
+    pub evaluators: Vec<String>,
```

Comment:
Not here because there aren't any evaluators that make sense to include, but perhaps down the line. Saying that, I don't think this will be super valuable, since users will likely want to augment that list (since the list of default evaluators won't be sufficient for any configuration, we'll have to add more ). In which case we could provide a variety of flags like `--base-evaluators`, `--additional-evaluators`, etc. but at that point it's getting more complex than it's worth.

Reply:
I think we're thinking of different user types here perhaps. I imagine that we and a few other highly proficient node operators will publish configuration files that users could feed into the NHC that they run. Indeed, the Helm chart I'm working on has config files bundled in, with appropriate values set for evaluators and each of their config options, for each different baseline config. The mid level user might use such a Helm chart / pre published config set to run NHC. However I expect most users won't go that far even and will just use the NHCs (likely via a web UI) that the first group of folks (us and other operators) host and have configured correctly.